base/abstractarray.jl

```diff
@@ -1152,12 +1152,30 @@ isempty(a::AbstractArray) = (length(a) == 0)
 
 ## range conversions ##
 
+map(::Type{T}, r::AbstractRange{T}) where {T} = r
 map(::Type{T}, r::StepRange) where {T<:Real} = T(r.start):T(r.step):T(last(r))
+map(::Type{T}, r::OneTo) where {T<:Integer} = oneto(T(last(r)))
+map(::Type{Bool}, r::OneTo) = map(Bool, UnitRange(r))
+map(::Type{T}, r::OneTo) where {T<:Real} = map(T, UnitRange(r))
 map(::Type{T}, r::UnitRange) where {T<:Real} = T(r.start):T(last(r))
 map(::Type{T}, r::StepRangeLen) where {T<:AbstractFloat} = convert(StepRangeLen{T}, r)
 function map(::Type{T}, r::LinRange) where T<:AbstractFloat
     LinRange(T(r.start), T(r.stop), length(r))
 end
+# Special conversions for IdentityUnitRange
+# Since IdentityUnitRange is defined as <: AbstractUnitRange{Int},
+# its eltype may not be converted in Base,
+# nor is it possible in general to map a type over it preserving its axes.
+# We define the `Int` case here so that OffsetArrays may define the conversion
+# for other eltypes through type-piracy.
+# In the future this type-piracy may not be necessary if IdentityUnitRange is
+# defined to be <: AbstractUnitRange{<:Integer}
+map(::Type{Int}, r::IdentityUnitRange) = r
+# For IdentityUnitRange{<:OneTo}, the map may be evaluated correctly in Base
+# We pop the parent range as the result may not always be representable as an IdentityUnitRange
+map(::Type{T}, r::IdentityUnitRange{<:OneTo}) where {T<:Real} = map(T, axes1(r))
+# this method is for ambiguity resolution
+map(::Type{Int}, r::IdentityUnitRange{<:OneTo}) = r
```

Comment:
For `r::IdentityUnitRange`, given that ` IdentityUnitRange(r.indices) === r`, we could possibly use the deconstruct-and-then-construct alternative to solve ambiguity:

```suggestion
Base.map(::Type{T}, r::IdentityUnitRange) where {T<:Real} = identity_or_map(T, axes1(r))
identity_or_map(::Type{Int}, r::AbstractUnitRange) = IdentityUnitRange(r) # the caller `map` thus becomes `identity`
identity_or_map(::Type{T}, r::AbstractUnitRange) = map(T, r)
```

Reply:
This will stack overflow as `axes1(r) === r` for `Base.IdentityUnitRange` if the parent range is not a `Base.OneTo`

```julia
julia> r = Base.IdentityUnitRange(2:3)
Base.IdentityUnitRange(2:3)

julia> map(Int32, r)
ERROR: StackOverflowError:
```